scipy/spatial/transform/_rotation_xp.py

```diff
@@ -0,0 +1,1138 @@
+        if vdot_ax1_ax2 >= 1e-7:
+            raise ValueError("Consecutive axes must be orthogonal.")
+
+    angles = _compute_davenport_from_quat(
+        quat, axes[0, ...], axes[1, ...], axes[2, ...], extrinsic
+    )
+    if degrees:
+        angles = _rad2deg(angles)
+    return angles
+
+
+def inv(quat: Array) -> Array:
+    xp = array_namespace(quat)
+    quat = xpx.at(quat)[..., :3].multiply(-1, copy=True, xp=xp)
+    return quat
+
+
+def magnitude(quat: Array) -> Array:
+    xp = array_namespace(quat)
+    sin_q = xp_vector_norm(quat[..., :3], axis=-1, xp=xp)
+    cos_q = xp.abs(quat[..., 3])
+    angles = 2 * xp.atan2(sin_q, cos_q)
+    return angles
+
+
+def approx_equal(
+    quat: Array, other: Array, atol: float | None = None, degrees: bool = False
+) -> Array:
+    xp = array_namespace(quat)
+    if atol is None:
+        if degrees:
+            warnings.warn(
+                "atol must be set to use the degrees flag, defaulting to 1e-8 radians.",
+                stacklevel=2,
+            )
+        atol = 1e-8
+    elif degrees:
+        atol = _deg2rad(atol)
+    atol = xp.asarray(atol, device=device(quat))
+
+    if not broadcastable(quat.shape, other.shape):
+        raise ValueError(
+            "Expected equal number of rotations in both "
+            "or a single rotation in either object, "
+            f"got {quat.shape[0]} rotations in first and {other.shape[0]} rotations in "
+            "second object."
+        )
+
+    quat_result = compose_quat(other, inv(quat))
+    angles = magnitude(quat_result)
+    return angles < atol
+
+
+def mean(quat: Array, weights: Array | None = None) -> Array:
+    xp = array_namespace(quat)
+    _device = device(quat)
+    dtype = xp_result_type(quat, force_floating=True, xp=xp)
+    if quat.shape[0] == 0:
+        raise ValueError("Mean of an empty rotation set is undefined.")
+
+    # Branching code is okay for checks that include meta info such as shapes and types
+    if weights is None:
+        weights = xp.ones(quat.shape[:-1], dtype=dtype, device=_device)
+        weights = xpx.atleast_nd(weights, ndim=1, xp=xp)
+    else:
+        weights = xp.asarray(weights, dtype=dtype, copy=True, device=_device)
+    if not is_lazy_array(weights) and xp.any(weights < 0):
+        raise ValueError("`weights` must be non-negative.")
+
+    # TODO: Missing full broadcasting support.
+    if weights.ndim != 1:
+        raise ValueError(
+            f"Expected `weights` to be 1 dimensional, got shape {weights.shape}."
+        )
+    n_rot = quat.shape[0] if quat.ndim > 1 else 1
+    if weights.shape[0] != n_rot:
+        raise ValueError(
+            "Expected `weights` to have number of values equal to number of rotations, "
+            f"got {weights.shape[0]} values and {n_rot} rotations."
+        )
+    # DECISION: We cannot check for negative weights because jit code needs to be
+    # non-branching. We return NaN instead
```

Comment:
The below pattern is different from everything above where you were branching for lazy/non-lazy. Is this deliberate?

Reply:
No. This seems to be a remnant from back when we discussed if lazy frameworks should have different behaviour from eager frameworks for error handling.